### PR TITLE
fix(next) - in next server actions, direct "use step" calls can fail in production because client-mode transform prunes step-only imports

### LINKED ISSUE #1311: Next server actions: direct "use step" calls can fail in production because client-mode transform prunes step-only imports
The docs say that calling a step outside a workflow should behave like a normal in-process function call:

https://useworkflow.dev/docs/foundations/workflows-and-steps#step-functions

"Calling a step from outside a workflow or from another step will essentially run the step in the same process like a normal function (in other words, the `use step` directive is a no-op)."

However, in prod build of NextJS with server actions, calling a step failed when it relies on imports.

In development it works.

In production, the direct step call fails at runtime because the compiled module keeps the step body but drops imports/local helpers that are only referenced from the step body.

So I got "ReferenceError: db is not defined"

Seems to come from the TransformMode in SWC - In `TransformMode::Client`, step bodies are kept callable, but usage analysis skips step-function bodies, so imports that are only used inside those step bodies can be removed.

Summary:
- Next.js App Router
- `use server`
- server action directly imports a `"use step"` function
- the step references imported helpers or imported runtime bindings e.g ```import db from packages/db```
- dev works, prod fails

I will open PR to fix + test soon

### PR DESCRIPTION
**Pull request by NathanColosimo**
Closes #1311 

This fixes a production-only bug where a `"use step"` function imported directly outside workflow context could fail because client-mode dead-code elimination removed imports/local helpers used only inside the step body.

## Root cause
`TransformMode::Client` keeps step bodies intact, but the usage collector treated step bodies like workflow-mode replacements and skipped analyzing them.

## Fix
In client mode, usage analysis now includes step-function bodies.

## Repro
Added a focused Next Turbopack repro for:
- App Router
- `use server`
- server action directly importing a `"use step"` function
- step depends on imported helpers

## Tests
Updated SWC fixture expectations for client-mode outputs where step-body-only imports/locals now need to be preserved.

`cargo test -p swc_workflow`

# Let me know if you want less tests / formatting I can make this more minimal!

### CODE REVIEW

File: packages/swc-plugin-workflow/transform/src/lib.rs [hunk 1]
```diff
@@ -3277,6 +3277,7 @@ impl StepTransform {
         let mut visitor = ComprehensiveUsageCollector {
             used_identifiers: &mut used_identifiers,
             step_function_names: &self.step_function_names,
+            analyze_step_function_bodies: self.mode == TransformMode::Client,
```
Comment (TooTallNate) [inline packages/swc-plugin-workflow/transform/src/lib.rs:3280]: **Blocking:** Per the repo's `AGENTS.md` instructions:

> When modifying the SWC compiler plugin (`packages/swc-plugin-workflow`), you must also update the specification document at `packages/swc-plugin-workflow/spec.md` to reflect any changes to the transformation behavior.

The spec currently has no mention of dead code elimination behavior. This PR changes DCE semantics (client mode now preserves step-body-only imports) and should document:
1. That DCE runs in both Workflow and Client modes
2. That in Client mode, step function bodies are preserved, so their dependencies (imports, local helpers) must also be preserved
3. That in Workflow mode, step function bodies are replaced with stubs, so body-only dependencies are eligible for removal

File: workbench/nextjs-webpack/app/home-client.tsx [hunk 2]
```diff
@@ -0,0 +1,383 @@
+'use client';
+
+import { useCallback, useEffect, useRef } from 'react';
+import { materializeWorkflowArgs } from '@/app/workflows/args';
```
Comment (TooTallNate) [inline workbench/nextjs-webpack/app/home-client.tsx:4]: **Blocking:** This file imports from `@/app/workflows/args` and `@/app/workflows/types`, but neither of those files exist in the `nextjs-webpack` workbench — only the `nextjs-turbopack` workbench gets those new files in this PR. Additionally, `workbench/nextjs-webpack/app/page.tsx` is not updated to use this new component.

This will break the webpack workbench. Either:
- Add the corresponding `types.ts`, `args.ts` files and update `page.tsx` in the webpack workbench too, or
- Remove this file from the PR if the webpack workbench refactoring isn't ready yet

File: packages/builders/src/base-builder.ts [hunk 3]
```diff
@@ -102,6 +102,7 @@ export abstract class BaseBuilder {
       '**/.git/**',
       '**/.next/**',
       '**/.nuxt/**',
+      '**/.output/**',
```
Comment (TooTallNate) [inline packages/builders/src/base-builder.ts:105]: Good catch adding `.output` to the ignore list. This prevents Nuxt/Nitro build artifacts from being picked up as input files.

File: packages/builders/src/base-builder.ts [hunk 4]
```diff
@@ -398,8 +399,10 @@ export abstract class BaseBuilder {
         /\\/g,
         '/'
       );
-      // Ensure relative paths start with ./ so esbuild resolves them correctly
-      if (!relativePath.startsWith('.')) {
+      // Ensure relative paths start with ./ so esbuild resolves them correctly.
+      // Paths like ".output/..." are not valid relative specifiers and must
+      // become "./.output/...".
```
Comment (TooTallNate) [inline packages/builders/src/base-builder.ts:404]: Good fix. The previous `!relativePath.startsWith('.')` would incorrectly match paths like `.output/server/index.mjs` (which start with `.` but are NOT valid relative specifiers). Checking for `./` and `../` specifically is the correct approach.

File: packages/builders/src/base-builder.ts [hunk 5]
```diff
@@ -879,9 +884,21 @@ export const POST = workflowEntrypoint(workflowCode);`;
       (f) => !inputFilesNormalized.has(f)
     );
 
-    // Re-exports for input files (user's workflow/step definitions)
+    // Re-exports for input files (user's workflow/step definitions).
+    // These must use valid relative specifiers because some frameworks pass
+    // generated files like ".output/server/index.mjs" as input files.
     const reexports = inputFiles
```
Comment (TooTallNate) [inline packages/builders/src/base-builder.ts:890]: Nit: The re-exports for `inputFiles` previously used the raw file paths directly (`export * from '${file}'`). The new relative path computation is more correct, but note that the same pattern should be applied consistently. The old code happened to work because `inputFiles` are absolute paths and esbuild resolves those, but relative paths are indeed better for portability.

File: packages/builders/src/discover-entries-esbuild-plugin.ts [hunk 6]
```diff
@@ -13,6 +13,17 @@ const enhancedResolve = promisify(enhancedResolveOriginal);
 
 export const jsTsRegex = /\.(ts|tsx|js|jsx|mjs|cjs|mts|cts)$/;
 
+function isGeneratedBuildArtifactPath(filePath: string): boolean {
```
Comment (TooTallNate) [inline packages/builders/src/discover-entries-esbuild-plugin.ts:16]: Nit: Consider whether this list should be kept in sync with the ignore patterns in `base-builder.ts` `getInputFiles()` (which also ignores `.next`, `.nuxt`, `.svelte-kit`, `.vercel`). Currently these are two independent lists. If a new framework output directory is added, it would need to be updated in both places. Could extract a shared constant.

File: lib/steps/paths-alias-test.ts [hunk 7]
```diff
@@ -4,6 +4,6 @@
  * Note: This is NOT a step function - it's a regular function that gets called
  * from within a step to verify path alias imports work correctly.
  */
-export function pathsAliasHelper(): string {
+export function pathsAliasHelper() {
```
Comment (TooTallNate) [inline lib/steps/paths-alias-test.ts:7]: Nit: This removes the explicit return type annotation `: string`. Was this intentional? The inferred type is the same, but explicit return types on exported functions can be helpful for API documentation and catching accidental changes.

File: workbench/nextjs-turbopack/app/workflows/definitions.ts [hunk 8]
```diff
@@ -1,12 +1,7 @@
-import { allWorkflows } from '@/_workflows';
+import 'server-only';
```
Comment (TooTallNate) [inline workbench/nextjs-turbopack/app/workflows/definitions.ts:1]: Nice — adding `server-only` here ensures `definitions.ts` (which imports `allWorkflows` containing actual workflow function references) can never accidentally be pulled into a client bundle. This is the correct architectural fix that motivates the `page.tsx` -> `home-client.tsx` split.

**Review by TooTallNate — DISMISSED** [review]
## Review Summary

The core SWC fix is correct and well-reasoned — client-mode DCE was incorrectly skipping step function body analysis, causing imports/helpers used only inside step bodies to be pruned. The `analyze_step_function_bodies` flag cleanly separates Workflow mode (bodies replaced → skip analysis) from Client mode (bodies preserved → must analyze). The test fixture updates confirm the fix works as expected.

The builder changes (relative path prefix fix, `.output` ignore, build artifact filtering) are also good and address real edge cases.

### Blocking Issues

1. **`spec.md` not updated** — Per repo rules (`AGENTS.md`), any change to the SWC plugin transform behavior must be reflected in `packages/swc-plugin-workflow/spec.md`. The spec currently has no mention of dead code elimination. This PR changes DCE semantics and should document the behavioral difference between Workflow and Client modes.

2. **`workbench/nextjs-webpack/app/home-client.tsx` is broken** — This file is added but imports from `@/app/workflows/args` and `@/app/workflows/types`, neither of which exist in the webpack workbench. The webpack `page.tsx` is also not updated to use the new component. This will cause build failures.

### Minor Notes
- The import path cleanup (removing `.js` extensions) and nuxt config fix are fine
- The repro page structure (`test-direct-step-call-server-action/`) follows Next.js patterns well
- Consider sharing the build artifact directory list between `base-builder.ts` and `discover-entries-esbuild-plugin.ts`

File: workbench/nextjs-webpack/app/home-client.tsx (hunk 2 above)
Comment (NathanColosimo) [reply workbench/nextjs-webpack/app/home-client.tsx:4]: fixed

File: packages/swc-plugin-workflow/transform/src/lib.rs (hunk 1 above)
Comment (NathanColosimo) [reply packages/swc-plugin-workflow/transform/src/lib.rs:3280]: added

File: lib/steps/paths-alias-test.ts (hunk 7 above)
Comment (NathanColosimo) [reply lib/steps/paths-alias-test.ts:7]: yes, was intentional, when it was there before was causing issues on webpack build? may want to look into this separately later

File: packages/builders/src/discover-entries-esbuild-plugin.ts (hunk 6 above)
Comment (NathanColosimo) [reply packages/builders/src/discover-entries-esbuild-plugin.ts:16]: future PR maybe

File: workbench/nextjs-webpack/app/home-client.tsx (hunk 2 above)
Comment (TooTallNate) [inline workbench/nextjs-webpack/app/home-client.tsx:4]: This file still imports from `@/app/workflows/args` and `@/app/workflows/types`, which do not exist in the webpack workbench. The webpack `page.tsx` and `definitions.ts` are also not updated. This will break the webpack workbench build.

Suggestion: remove this file from the PR and do the webpack workbench refactoring in a follow-up.

File: packages/swc-plugin-workflow/spec.md [hunk 9]
```diff
@@ -412,6 +412,8 @@ registerStepFunction("step//./input//subtract", subtract);
 
 In workflow mode, step function bodies are replaced with a `globalThis[Symbol.for("WORKFLOW_USE_STEP")]` call. Workflow functions keep their bodies and are registered with `globalThis.__private_workflows.set()`.
 
+After the workflow-mode rewrite, the transform also runs a dead code elimination (DCE) pass. This pruning only affects the emitted workflow/client outputs, not step-mode output. In workflow mode, because step bodies are replaced with step proxies, imports, helper functions, nested steps, and other pure statements that were only referenced from those original step bodies become eligible for removal. Exports and any identifiers still referenced by the transformed workflow code are preserved.
```
Comment (TooTallNate) [inline packages/swc-plugin-workflow/spec.md:415]: This is a clear and accurate description of the workflow-mode DCE behavior. The distinction between what gets pruned (step-body-only references) vs. what's preserved (exports and identifiers still referenced by transformed code) is well stated.

**Review by TooTallNate — DISMISSED** [review]
## Re-review

Thanks for the updates. The `spec.md` addition is well-written and accurately documents the DCE behavioral difference between Workflow and Client modes — that blocking issue is resolved.

However, the **webpack workbench issue is still present**. The only file changed in `workbench/nextjs-webpack/` is `app/home-client.tsx`, which imports from:
- `@/app/workflows/args` — does not exist in the webpack workbench
- `@/app/workflows/types` — does not exist in the webpack workbench

Additionally, `workbench/nextjs-webpack/app/page.tsx` is not updated to use the new `HomeClient` component, and `workbench/nextjs-webpack/app/workflows/definitions.ts` still has the old structure (no `server-only`, still exports `materializeWorkflowArgs` and `WorkflowDefinition` directly).

The simplest fix would be to **remove `workbench/nextjs-webpack/app/home-client.tsx` from this PR** entirely, since the webpack workbench refactoring is incomplete. That can be done in a follow-up.

File: workbench/nextjs-webpack/app/home-client.tsx (hunk 2 above)
Comment (NathanColosimo) [reply workbench/nextjs-webpack/app/home-client.tsx:4]: These files already did exist

image

**Comment by NathanColosimo** [comment]
@TooTallNate The files did already exist

File: workbench/nextjs-webpack/app/home-client.tsx [hunk 10]
```diff
@@ -0,0 +1,383 @@
+'use client';
```
Comment (TooTallNate) [inline workbench/nextjs-webpack/app/home-client.tsx:1]: Nit: Every other `.tsx` file in `workbench/nextjs-webpack/app/` is a symlink to the turbopack workbench (`page.tsx`, `layout.tsx`, `globals.css`). This file is added as a real file with identical content. For consistency, consider making it a symlink too:
```
ln -s ../../nextjs-turbopack/app/home-client.tsx workbench/nextjs-webpack/app/home-client.tsx
```
Not blocking — it works either way.
